File: NumberOfIslands/NumberOfIslands.py
class Solution:
    grid = [[None]]
    n = 0
    m = 0
    
    def island(self, i, j):
        if i > 0 and self.grid[i-1][j] == '1':
            self.grid[i-1][j] = '2'
            self.island(i-1, j)
        if i < self.n-1 and self.grid[i+1][j] == '1':
            self.grid[i+1][j] = '2'
            self.island(i+1, j)
        if j > 0 and self.grid[i][j-1] == '1':
            self.grid[i][j-1] = '2'
            self.island(i, j-1)
        if j < self.m-1 and self.grid[i][j+1] == '1':
            self.grid[i][j+1] = '2'
            self.island(i, j+1)
        
    
    def numIslands(self, grid: List[List[str]]) -> int:
        self.grid = grid
        self.n = len(grid)
        if grid != []:
            self.m = len(grid[0])
        else:
            return 0
        
        cnt = 0
        
        for i in range(self.n):
            for j in range(self.m):
                if self.grid[i][j] == '1':
                    self.island(i, j)
                    cnt += 1

File: code/NumberOfIslands/NumberOfIslands.cpp
struct Coord {
    int i, j;
};

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        // BFS
        queue<Coord> q;

        int island = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') { // initiate a new BFS
                    q.push({i, j});

                    while (!q.empty()) {
                        Coord p = q.front();
                        q.pop();

                        if (p.i >= 0 && p.i < grid.size() && p.j >= 0 &&
                            p.j < grid[0].size() && grid[p.i][p.j] == '1') {
                            grid[p.i][p.j] = '0';
                            q.push({p.i + 1, p.j});
                            q.push({p.i - 1, p.j});
                            q.push({p.i, p.j + 1});
                            q.push({p.i, p.j - 1});
                        }
                    }

                    island++;
                }
            }
        }

        return island;
    }
};
